File: test_feature/hr/impl/src/main/java/com/github/klee0kai/stone/test_feature/hr/di/DepartmentsModule.java
package com.github.klee0kai.stone.test_feature.hr.di;

import com.github.klee0kai.stone.annotations.module.Module;
import com.github.klee0kai.stone.annotations.module.Provide;
import com.github.klee0kai.stone.test_feature.hr.department.SecurityDepartment;
import com.github.klee0kai.stone.test_feature.hr.department.SecurityDepartmentImpl;
import com.github.klee0kai.stone.test_feature.hr.department.SoftwareDepartment;
import com.github.klee0kai.stone.test_feature.hr.department.SoftwareDepartmentImpl;
import com.github.klee0kai.stone.test_feature.hr.store.EmployeesStore;

@Module
public class DepartmentsModule {

    @Provide(cache = Provide.CacheType.Strong)
    public EmployeesStore employeesStore() {
        return new EmployeesStore();
    }


    @Provide(cache = Provide.CacheType.Soft)
    public SecurityDepartment securityDepartment() {
        return new SecurityDepartmentImpl();
    }

    @Provide(cache = Provide.CacheType.Soft)
    public SoftwareDepartment softwareDepartment() {
        return new SoftwareDepartmentImpl();
    }

}

File: test_feature/hr/impl/src/main/java/com/github/klee0kai/stone/test_feature/hr/di/DepartmentsModule.kt
package com.github.klee0kai.stone.test_feature.hr.di

import com.github.klee0kai.stone.annotations.module.Module
import com.github.klee0kai.stone.annotations.module.Provide
import com.github.klee0kai.stone.test_feature.hr.department.SecurityDepartment
import com.github.klee0kai.stone.test_feature.hr.department.SecurityDepartmentImpl
import com.github.klee0kai.stone.test_feature.hr.department.SoftwareDepartment
import com.github.klee0kai.stone.test_feature.hr.department.SoftwareDepartmentImpl
import com.github.klee0kai.stone.test_feature.hr.store.EmployeesStore

@Module
interface DepartmentsModule {

    @Provide(cache = Provide.CacheType.Strong)
    fun employeesStore(): EmployeesStore {
        return EmployeesStore()
    }


    @Provide(cache = Provide.CacheType.Soft)
    fun securityDepartment(): SecurityDepartment {
        return SecurityDepartmentImpl()
    }

    @Provide(cache = Provide.CacheType.Soft)
    fun softwareDepartment(): SoftwareDepartment {
        return SoftwareDepartmentImpl()
    }
}
